Do not imply having parsed the request body

Zegnat opened the pull request:
Strictly follow the PSR-7 guidance where `$_POST` is used for a request’s parsed body only for very specific requests. All other requests should result in a null value and leave other parsing up to application logic.

The current practice of putting an empty array in may imply we have done parsing and found an empty request body. See also Nyholm/psr7#100.

yaskhan commented on src/ServerRequestCreator.php:
```diff
@@ -51,13 +51,18 @@ public function fromGlobals(): ServerRequestInterface
 
-        return $this->fromArrays($server, $headers, $_COOKIE, $_GET, $_POST, $_FILES, \fopen('php://input', 'r') ?: null);
+        $post = null;
+        if ('POST' === $this->getMethodFromEnv($server) && \in_array($headers['content-type'], ['application/x-www-form-urlencoded', 'multipart/form-data'])) {
```
I think that code not case insensitive and not support header values like `application/x-www-form-urlencoded;charset= charset=utf-8`. And it seems not enough `application/json`

Zegnat replied:
Are MIME types case insensitive? Maybe they are, I never actually considered. Good catch. Will do some research and patch accordingly. Should definitely support the `;charset` case!

`application/json` is not a POST body that PHP natively parses, so it is not something we can trigger `withParsedBody` for. I would recommend using some sort of middleware (e.g. [Payload](https://github.com/middlewares/payload)) if you need special request parsing.

Zegnat commented:
@yaskhan thanks again for the review! I think I have now addressed your points, as well as added some tests to try and keep future contributions from breaking anything!

Now I just need to figure out why styleci is breaking even after running `php-cs-fixer`. Are the configurations correct, @Nyholm?

Zegnat commented:
Turns out [StyleCI’s Symfony presets](https://docs.styleci.io/presets#symfony) include a bunch of “blank line before” rules, while PHP CS Fixer actually disables all of them except for before `return` when you follow the Symfony presets.

Changed the rules to match StyleCI. @Nyholm let me know if you’d rather change StyleCI settings than PHP CS Fixer settings. I am not sure which one is seen as leading here—or by Symfony.

This will always be BC breaking. Although on the parameter side I merely widened the allowed input, which in itself will not break BC (and is in line with LSP), there may be minor breakage in the output as `ServerRequestInterface::getParsedBody()` may now return `null` where it previously would’ve returned an empty array.

Although this is a BC break from the PSR7 Server perspective, the `ServerRequestInterface` always allowed for `null` to be returned there so I am not sure this is likely to affect anyone. And we aren’t on a stable 1.0.0 semver API anyway.

Nyholm commented on src/ServerRequestCreator.php:
```diff
@@ -51,13 +51,30 @@ public function fromGlobals(): ServerRequestInterface
+                    \strtolower(\trim(\explode(';', $headerValue, 2)[0])),
+                    ['application/x-www-form-urlencoded', 'multipart/form-data']
+                )) {
+                    $post = $_POST;
```
What will $_POST contain if it isnt a POST request with content type `application/x-www-form-urlencoded or multipart/form-data`?

Nyholm replied:
It is just an empty array.

Zegnat replied:
Yes. From my reading of PSR-7, `getParsedBody` should return `null` when no (parseable) request body was provided. If we pass along the empty array from `$_POST` that would be the same as saying we successfully parsed the body content, which may not have been try at al. E.g. we make no attempt to parse JSON bodies.

That is why I wanted to establish a clear difference between `[]` and `null` 😄

Nyholm replied:
Thank you

Nyholm commented:
Thank you
